test/semantics/dosemantics09.f90

```diff
@@ -0,0 +1,125 @@
+! Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+!
+! Licensed under the Apache License, Version 2.0 (the "License");
+! you may not use this file except in compliance with the License.
+! You may obtain a copy of the License at
+!
+!     http://www.apache.org/licenses/LICENSE-2.0
+!
+! Unless required by applicable law or agreed to in writing, software
+! distributed under the License is distributed on an "AS IS" BASIS,
+! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+! See the License for the specific language governing permissions and
+! limitations under the License.
+
+!C1129 
+!A variable that is referenced by the scalar-mask-expr of a
+!concurrent-header or by any concurrent-limit or concurrent-step in that
+!concurrent-header shall not appear in a LOCAL locality-spec in the same DO
+!CONCURRENT statement.
+
+subroutine s1()
+
+!ERROR: 'i' is already declared in this scoping unit
+  do concurrent (i=1:10) local(i)
+  end do
+end subroutine s1
+
+subroutine s2()
+!ERROR: 'i' is already declared in this scoping unit
+  do concurrent (i=1:10) local_init(i)
+  end do
+end subroutine s2
+
+subroutine s3()
+!ERROR: 'i' is already declared in this scoping unit
+  do concurrent (i=i:10) shared(i)
+  end do
+end subroutine s3
+
+subroutine s4()
+!ERROR: concurrent-header expression references variable 'i' in locality-spec
+  do concurrent (j=i:10) local(i)
```

Comment:
This message points to the expression where `i` is referenced but not the locality spec that causes the error. I think it would be easier to understand if it did, i.e. have the attached message point to the `i` in `local(i)`.

Or have the error point to the second `i` and the attached message point to the first one. That would require rewording the message, something like this:
`'i' may not appear in LOCAL locality-spec because it is referenced in the concurrent-header`

Also, this constraint only applies to the LOCAL locality-spec so I think the message should say that.

Reply:
I like the suggestion of specifying that the problem is specific to LOCAL locality-specs.  But I think it's better for the error message to refer to the variable name's use in the expression since I'd guess that this location would be more difficult to find than the name in the locality-spec.  Also, referring to the name's declaration is a good thing since the actual declaration might require traversing a construct association which changes the name of the variable.

I'll change the text of the error message.